### ISSUE TITLE
Audio Mute not working for Amazon Prime Video

### ISSUE THREAD

**Issue by jsimonrichard** [richardfrost/AdvancedProfanityFilter#325]
#### :bug: Description
As far as I can tell, this extension is supposed to work on Amazon (I'm assuming that means Prime Video). However, ever with subtitles on, the audio is never muted. The subtitle text, however, is censored.

#### :twisted_rightwards_arrows: Steps To Reproduce
1. Go to https://www.amazon.com/Amazon-Video/b?ie=UTF8&node=2858778011
2. Click on any video with bad language
3. Turn on subtitles

#### :heavy_check_mark: Expected behavior
I expect the tab to get muted when there is profanity in the subtitles.

#### :computer: Details
 - **Device:** Systems76
 - **OS:** Pop!_OS
 - **Browser:** firefox
 - **Browser Version:** 90.0
 - **Affected site(s):** www.amazon.com

**Comment by richardfrost** [comment]
Thanks for reaching out about this issue! Sorry its not working for you. You are correct that it is intended to work on Amazon Prime Video titles. Just a couple things to check first off:

- Do you have audio muting enabled in the extension? (Options page > Audio tab)
- Have you tried it on another site, such as YouTube? This helps to confirm that the audio muting is active and working as intended, and will help us isolate the issue.

By default the filter tries to filter all text on the page, and the captions/subtitles fall into that category. The audio muting function is there to identify which text on the screen is a caption/subtitle vs other text.

**Comment by jsimonrichard** [comment]
Audio muting is enabled, and it does work on YouTube.

Also, I'm using the "Hide Captions" option, so I can tell when the extension has found the subtitle element. The captions on YouTube are hidden, but the subtitles on Amazon Prime are not.

**Comment by richardfrost** [comment]
Thanks for the info. Its possible Amazon has updated/changed something that has broken the filter's ability to identify the captions/subtitles. Can you share an example or two of videos that it is not working on?

**Comment by jsimonrichard** [comment]
I don't think it's working on any of the videos, but the TV show "The Boys" is a good example. There's a lot of swearing in that one.

**Comment by richardfrost** [comment]
Okay, I was able to just test it, and it seemed to work for me. The Boys S01E01 @ 0:00:40 (probably go back a few seconds before that), there is an example (using the default word list). Which mute method are you using (Extension Option page > Audio tab)? Could you try switching it to the other? What language of captions/subtitles are you using?

If its still not working, a couple other things that could help in troubleshooting: Can you pause the video while captions are being shown, and then right-click on them and choose "Inspect element", and then take a screenshot like I have below:

![image](https://user-images.githubusercontent.com/1507205/128575180-9b422afd-18c4-4138-b92b-25c83e939b9b.png)

The other thing would be to attach your config here (Extension Options page > Config tab > Export button).

**Comment by jsimonrichard** [comment]
I was using the Mute Tab method, but I also tried the Video Volume method and it didn't work. Also, FYI, I'm using English subtitles.

Here's a screenshot of the inspect element.
![Screenshot from 2021-08-06 20-38-40](https://user-images.githubusercontent.com/15972974/128582333-38b1fb53-fca9-4b9a-bc21-588ad1aebc46.png)

Here's the config file, excluding the word list.
```json
{
  "audioWordlistId": 0,
  "censorCharacter": "*",
  "censorFixedLength": 0,
  "collectStats": true,
  "customAudioSites": {
  },
  "darkMode": false,
  "defaultSubstitution": "censored",
  "defaultWordMatchMethod": 0,
  "defaultWordRepeat": 0,
  "defaultWordSeparators": 0,
  "domains": {
    "www.amazon.com": {
      "enabled": true,
      "wordlist": 0,
      "audioList": 0
    },
    "www.disneyplus.com": {
      "audioList": 0,
      "enabled": true,
      "wordlist": 0
    },
    "www.hulu.com": {
      "enabled": true
    },
    "www.youtube.com": {
      "wordlist": 0,
      "audioList": 0,
      "enabled": true
    }
  },
  "enabledDomainsOnly": true,
  "fillerAudio": "",
  "filterMethod": 0,
  "filterWordList": true,
  "iWordWhitelist": [],
  "muteAudio": true,
  "muteAudioOnly": false,
  "muteCueRequireShowing": false,
  "muteMethod": 0,
  "password": null,
  "preserveCase": true,
  "preserveFirst": true,
  "preserveLast": false,
  "showCounter": true,
  "showSubtitles": 3,
  "showSummary": true,
  "showUpdateNotification": true,
  "substitutionMark": false,
  "wordWhitelist": [],
  "wordlistId": 0,
  "wordlists": [
    "Wordlist 1",
    "Wordlist 2",
    "Wordlist 3",
    "Wordlist 4",
    "Wordlist 5",
    "Wordlist 6"
  ],
  "wordlistsEnabled": true,
  "words": { ... },
  "youTubeAutoSubsMax": 0,
  "youTubeAutoSubsMin": 0
}
```

**Comment by richardfrost** [comment]
That's very interesting. If I had to guess, either your browser or OS is missing a codec, or DRM mechanism and Amazon is falling back to an older/more compatible player than what I get on any of my browsers. I don't currently have a linux box with a GUI, but I can probably test a live CD later if that's needed.

Based on the DOM elements that I can see in your screenshot, can you try adding this to your Custom Sites (Extension Option page > Audio tab), and then click save underneath the box:

```json
{
  "www.amazon.com": [
    {
      "displayHide": "none",
      "displaySelector": "div.controlsOverlay div.persistentPanel p > span",
      "displayShow": "inline-block",
      "iframe": false,
      "mode": "watcher",
      "parentSelector": "div.controlsOverlay div.persistentPanel p > span",
      "subtitleSelector": "div.controlsOverlay div.persistentPanel p > span > span",
      "videoSelector": "video"
    }
  ]
}
```

If that still doesn't work, could you get me a screen shot like you did before, but this time have the subtitles at the bottom of the screen so I can see more of what's going on above?

**Comment by jsimonrichard** [comment]
I wasn't able to get it working on Firefox.

However, even without these settings, I was able to get it working on Chromium. Maybe firefox is missing a codec or DRM mechanism or something. I can't get the subtitles to hide on Chromium, though.

**Comment by richardfrost** [comment]
Okay, so probably not your OS then, but for whatever reason that build of Firefox must be missing something. That's interesting about the captions.. so they are showing all the time in Chromium?

**Comment by jsimonrichard** [comment]
Correct.

**Comment by richardfrost** [comment]
Sorry for the delay on this... I got swamped with other things and this fell by the wayside. Are you still having the issue with the captions text not hiding in Chromium?

Or if you aren't interested in troubleshooting this anymore that's fine and we can close this. Thanks for helping to make the extension better.

**Comment by jsimonrichard** [comment]
Yes, I'm still having issues with the captions. Let me know what I can do to help.

**Comment by richardfrost** [comment]
@jsimonrichard I'm sorry, it looks like I let this one fall through the cracks. Are you still interested in working on this issue?

**Comment by jsimonrichard** [comment]
Sure. Let me know what you need

**Comment by richardfrost** [comment]
Awesome! So, just to recap where we were before. You tested on both Firefox and a chromium browser, right? And it sounds like on the chromium browser muting worked, but captions/subtitles couldn't be hidden. On Firefox, neither hiding or muting worked?

Just want to make sure we are on the same page, and that things haven't changed since then. 

Also, for troubleshooting purposes, could you make a backup of your APF settings (Options page > Config tab > Export), and then restore the defaults (same page)? Once that's done, go to the Audio tab and enable audio muting. Then, head back over to Amazon and test the video again? If you'd rather not test with a profane word, feel free to add an inoffensive word that shows up in the captions temporarily to the filter.

**Comment by jsimonrichard** [comment]
I tested it again with the default settings. It's the same as before:
* It works on chromium, but the subtitles are still shown (even with the **Hide all** option selected)
* It does not work at all on firefox

**Comment by richardfrost** [comment]
Thanks for testing again. Sorry for the delay.. I had to setup a VM to try and figure out what was going on. It appears that Amazon is serving a different page structure when I'm on Firefox in Pop!_OS. I'm working on Firefox first. Could you try this custom audio config on Firefox?

{
  "www.amazon.com": [
    {
      "displayHide": "none",
      "displaySelector": "div.webPlayerContainer div p > span",
      "displayShow": "",
      "iframe": false,
      "mode": "watcher",
      "muteMethod": 1,
      "parentSelector": "div.webPlayerContainer div span > span",
      "videoSelector": "div.webPlayerElement video[src]",
    }
  ]
}

**Comment by jsimonrichard** [comment]
This basically worked (selecting **hide all** hides the captions), but when the captions change they're shown for a fraction of a second before they're hidden again.

**Comment by richardfrost** [comment]
Yeah, unfortunately that flickering is pretty commonplace with Amazon. Basically it boils down to the fact that the extension is watching the page every ~20ms, so its possible to get that flickering of the text briefly. It happens, though isn't quite as bad in my experience on Chrome. Most other services don't have this issue because it can watch mutations, which happen before the text elements get displayed.

Does that config in Chromium happen to work about the same?

**Comment by jsimonrichard** [comment]
Yeah, it works basically the same.

By the way (I forgot to mention this earlier), but your JSON is malformed. There's an extra comma at the end of the last key/value line. I know it's not a big deal, but APF won't let you submit the custom configuration form until it's fixed so I thought I'd mention it before someone else tries to copy and paste it.

**Comment by richardfrost** [comment]
Thanks for the heads up. I'm usually better at actually testing it, but was in a hurry. Oops. I've updated it now so it should actually be valid.

It sounds like we've got it working up to par with other operating systems, so that's something. Hopefully I'll get some time to improve it even more sometime in the future. For now, I think I'll go ahead and close this issue, and add a new wiki page to store alternate site configurations and link it here, just in case we need to make exceptions like this in other cases.
